File: backend/assetflow-company-service/src/main/java/com/assetflow/company/controller/CompanyController.java
package com.assetflow.company.controller;

import com.assetflow.company.response.ApiResponse;
import com.assetflow.company.service.CompanyDataService;
import io.jsonwebtoken.Claims;
import java.util.*;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;
import org.springframework.http.HttpStatus;
import org.springframework.web.server.ResponseStatusException;

@RestController
public class CompanyController {
	private final CompanyDataService service;

	public CompanyController(CompanyDataService s) {
		service = s;
	}

	private Long company(Authentication a) {
		return ((Claims) a.getPrincipal()).get("companyId", Long.class);
	}
	private Long user(Authentication a) {
		return Long.valueOf(((Claims) a.getPrincipal()).getSubject());
	}
	private String role(Authentication a) {
		return ((Claims) a.getPrincipal()).get("role", String.class);
	}
	private void assertAdmin(Authentication a) {
		String role = ((Claims) a.getPrincipal()).get("role", String.class);
		if (!"SUPER_ADMIN".equals(role) && !"COMPANY_ADMIN".equals(role))
			throw new ResponseStatusException(HttpStatus.FORBIDDEN, "Administrator access is required");
	}

	@GetMapping("/api/{type:department|employee|admin|location}")
	public ApiResponse<List<Map<String, Object>>> list(@PathVariable String type,
			@RequestParam(required = false) String search, @RequestParam(defaultValue = "0") int page,
			@RequestParam(defaultValue = "25") int size, Authentication a) {
		return ApiResponse.ok(service.list(type, company(a), user(a), role(a), search, page, size));
	}

	@GetMapping("/api/{type:department|employee|admin|location}/{id}")
	public ApiResponse<Map<String, Object>> one(@PathVariable String type, @PathVariable Long id, Authentication a) {
		return ApiResponse.ok(service.one(type, company(a), user(a), role(a), id));
	}

	@PostMapping("/api/{type:department|employee|admin|location}")
	public ApiResponse<Map<String, Object>> create(@PathVariable String type, @RequestBody Map<String, Object> b,
			Authentication a) {
		assertAdmin(a);
		return ApiResponse.ok(service.create(type, company(a), user(a), b));
	}
	@PostMapping("/api/employee/import")
	public ApiResponse<Map<String, Object>> importEmployees(@RequestBody List<Map<String, Object>> rows, Authentication a) {
		assertAdmin(a);
		return ApiResponse.ok(service.importEmployees(company(a), user(a), rows));
	}
	@PutMapping("/api/{type:department|employee|admin|location}/{id}")
	public ApiResponse<Map<String,Object>> update(@PathVariable String type, @PathVariable Long id, @RequestBody Map<String,Object> b, Authentication a) {
		assertAdmin(a);
		return ApiResponse.ok(service.update(type, company(a), user(a), id, b));
	}

	@DeleteMapping("/api/{type:department|employee|admin|location}/{id}")
	public ApiResponse<Void> remove(@PathVariable String type, @PathVariable Long id, Authentication a) {
		assertAdmin(a);
		service.delete(type, company(a), user(a), id);
		return ApiResponse.ok(null);
	}
}

File: dotnet-backend/Assetflow.CompanyService/Controllers/CompanyController.cs
using System.Security.Claims;
using Assetflow.CompanyService.Models;
using Assetflow.CompanyService.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Assetflow.CompanyService.Controllers
{
    [ApiController]
    [Authorize]
    public class CompanyController : ControllerBase
    {
        private readonly CompanyDataService _service;

        public CompanyController(CompanyDataService service)
        {
            _service = service;
        }

        private long CompanyId()
        {
            var claim = User.FindFirst("companyId")?.Value;
            return claim != null ? long.Parse(claim) : 0;
        }

        private long UserId()
        {
            var claim = User.FindFirst(ClaimTypes.NameIdentifier) ?? User.FindFirst("sub");
            return claim != null ? long.Parse(claim.Value) : 0;
        }

        private string Role()
        {
            return User.FindFirst("role")?.Value ?? "";
        }

        private void AssertAdmin()
        {
            var role = Role();
            if (role != "SUPER_ADMIN" && role != "COMPANY_ADMIN")
            {
                throw new UnauthorizedAccessException("Administrator access is required");
            }
        }

        private bool IsValidType(string type)
        {
            return type is "department" or "employee" or "admin" or "location";
        }

        [HttpGet("api/{type}")]
        public async Task<ActionResult<ApiResponse<IEnumerable<IDictionary<string, object>>>>> List(
            string type,
            [FromQuery] string? search,
            [FromQuery] int page = 0,
            [FromQuery] int size = 25)
        {
            if (!IsValidType(type)) return NotFound();
            var data = await _service.ListAsync(type, CompanyId(), UserId(), Role(), search, page, size);
            return Ok(ApiResponse<IEnumerable<IDictionary<string, object>>>.Ok(data));
        }

        [HttpGet("api/{type}/{id}")]
        public async Task<ActionResult<ApiResponse<IDictionary<string, object>>>> One(string type, long id)
        {
            if (!IsValidType(type)) return NotFound();
            var data = await _service.OneAsync(type, CompanyId(), UserId(), Role(), id);
            return Ok(ApiResponse<IDictionary<string, object>>.Ok(data));
        }

        [HttpPost("api/{type}")]
        public async Task<ActionResult<ApiResponse<IDictionary<string, object>>>> Create(string type, [FromBody] IDictionary<string, object> body)
        {
            if (!IsValidType(type)) return NotFound();
            AssertAdmin();
            var data = await _service.CreateAsync(type, CompanyId(), UserId(), body);
            return Ok(ApiResponse<IDictionary<string, object>>.Ok(data));
        }

        [HttpPost("api/employee/import")]
        public async Task<ActionResult<ApiResponse<IDictionary<string, object>>>> ImportEmployees([FromBody] IEnumerable<IDictionary<string, object>> rows)
        {
            AssertAdmin();
            var data = await _service.ImportEmployeesAsync(CompanyId(), UserId(), rows);
            return Ok(ApiResponse<IDictionary<string, object>>.Ok(data));
        }

        [HttpPut("api/{type}/{id}")]
        public async Task<ActionResult<ApiResponse<IDictionary<string, object>>>> Update(string type, long id, [FromBody] IDictionary<string, object> body)
        {
            if (!IsValidType(type)) return NotFound();
            AssertAdmin();
            var data = await _service.UpdateAsync(type, CompanyId(), UserId(), id, body);
            return Ok(ApiResponse<IDictionary<string, object>>.Ok(data));
        }

        [HttpDelete("api/{type}/{id}")]
        public async Task<ActionResult<ApiResponse<object>>> Delete(string type, long id)
        {
            if (!IsValidType(type)) return NotFound();
            AssertAdmin();
            await _service.DeleteAsync(type, CompanyId(), UserId(), id);
            return Ok(ApiResponse<object>.Ok(null!));
        }
    }
}
